### PR TITLE
getzig: Add script to download zig toolchain

### PR DESCRIPTION
**Pull request by camh-**
Add a script to download zig releases, similar to `getgo`. It installs
it in `/opt/zig/<version>` and symlinks `/opt/zig/{stable,master}` to it
(depending on whether you have opted to download the latest stable or
master release.

### CODE REVIEW

File: getzig [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+*)
+	printf 'Unknown OS name: %s\n' "$(uname)" >&2
+	exit 1
+	;;
+esac
+
+case $(uname -m) in
+x86_64) ARCH=x86_64 ;;
+aarch64) ARCH=aarch64 ;;
+arm64) ARCH=aarch64 ;;
```
Comment (camh-) [inline getzig]: Hmmm, I should combine these:
```
aarch64|arm64) ARCH=aarch64 ;;
```

File: getzig [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+wetrun() {
+	if ((GETZIG_DEBUG > 0 || GETZIG_DRYRUN > 0)); then
+		echo "$@" >&2
+	fi
+	"$@"
+}
+
+# debug sends stdin to stdout, and if debug mode is on, also sends stdin to
+# stderr.
+debug() {
```
Comment (camh-) [inline getzig:127]: I think I should rename this to `dumpstdin` or something

File: getzig [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+	if ((GETZIG_DEBUG > 0)); then
+		tee /dev/stderr
+	else
+		cat
+	fi
+}
+#-----------------------------------------------------------------------------
+[[ "$(caller)" != 0\ * ]] || main "$@"
```
Comment (camh-) [inline getzig]: I should change this to my new construct: 
```
# Only run main if executed as a script and not sourced.
if [[ "${BASH_SOURCE[0]}" == "$0" ]]; then main "$@"; fi
```

**Review by juliaogris — COMMENTED** [review]
I think you should add `shfmt` to make / CI.
Maybe also a note for deps on newer `bash` / `jq` / `curl` ?
With bad input I get: 
```
./getzig: line 40: info["${key}"]: bad array subscript`
```

File: getzig (hunk 1 above)
Comment (juliaogris) [reply getzig]: sounds reasonable.

File: getzig [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+		# may not be the case with gojq
+		select_release='to_entries | .[1] | {version: .key} + .value'
+		;;
+	*)
+		printf 'unknown release type. Must be stable (default) or master\n' >&2
+		exit 1
+		;;
+	esac
+
+	wetrun curl -sL "https://ziglang.org/download/index.json" |
```
Comment (juliaogris) [inline getzig]: 🍭

Comment (juliaogris) [reply getzig]: to me wetrun is a confusing term. alwaysrun or forcerun would make more sense to me.
(you choose dry, so i would have expected that you chose wet)...
feel free to ignore if this is a common pattern / doesn't resonate.

File: getzig (hunk 2 above)
Comment (juliaogris) [reply getzig:127]: debug makes sense to me, lines up with GETZIG_DEBUG

File: getzig (hunk 3 above)
Comment (juliaogris) [reply getzig]: sounds good.

File: getzig (hunk 1 above)
Comment (camh-) [reply getzig]: done

File: getzig (hunk 2 above)
Comment (camh-) [reply getzig:127]: ok. i left it as debug

File: getzig (hunk 3 above)
Comment (camh-) [reply getzig]: done

File: getzig (hunk 4 above)
Comment (camh-) [reply getzig]: ok. i changed it to `alwaysrun`

**Comment by camh-** [comment]
> Maybe also a note for deps on newer `bash` / `jq` / `curl` ?

I've added a check for all these deps so it give a useful error.

> With bad input I get: [...]

Yeah, that was a bug. I've fixed it I think. The `set -e` was not taking effect because of where `get_latest_info` was run. I've moved that so it is returns an error, the script will error instead of continuing and operating on bad data.

> I think you should add shfmt to make / CI.

Ok. But not this PR. I would need to reformat getgo first/concurrently.

**Comment by camh-** [comment]
I moved the os/arch stuff into `get_latest_info` since that's the only place they were used. Now everything is in functions as it should be.

**Review by juliaogris — APPROVED** [review]
Lgtm, but I only re-reviewed on phone. If you want I try running it again on my Mac tomorrow.
🟩

**Pull request merged**